File: leetcode/0001-1000/0401-0500/0443-string-compression/StringCompression.java
public class StringCompression {
  public static void main(String[] args) {
    System.out.println(new Solution().compress(new char[] { 'a', 'a', 'b', 'b', 'c', 'c', 'c' }));
  }
}

class Solution {
  public int compress(char[] chars) {
    StringBuilder sb = new StringBuilder();

    for (int i = 1, streak = 1; i <= chars.length; i++) {
      if (i < chars.length && chars[i] == chars[i - 1]) {
        streak++;

        continue;
      }

      sb.append(chars[i - 1]);

      if (streak > 1) {
        sb.append(streak);
      }

      streak = 1;
    }

    sb.getChars(0, sb.length(), chars, 0);

    return sb.length();
  }
}

File: leetcode/0001-1000/0401-0500/0441-0450/0443-string-compression/StringCompression.ts
function compress(chars: string[]): number {
  const length = chars.length;

  let i = 0;
  let w = 0;

  while (i < length) {
    let j = i + 1;

    while (j < length && chars[j] === chars[i]) {
      j++;
    }

    chars[w++] = chars[i];

    if (j - i > 1) {
      `${j - i}`.split("").forEach(
        c => chars[w++] = c
      );
    }

    i = j;
  }

  return w;
};

console.log(compress(["a", "a", "b", "b", "c", "c", "c"]));
